webapp/channels/src/packages/mattermost-redux/src/reducers/entities/channel_categories.ts

```diff
@@ -92,7 +93,38 @@ export function byId(state: IDMappedObjects<ChannelCategory> = {}, action: Gener
 
         return changed ? nextState : state;
     }
+    case ChannelTypes.GM_CONVERTED_TO_CHANNEL: {
+        // For GM to Private channel conversion feature
+        // In the case when someone converts your GM to a private channel and moves it to a team
+        // you're not currently on, we need to remove the channel from "direct messages" category
+        // and add it to "channels" category of target team. Even though the server sends a websocket event about updated category data,
+        // it does so only for the team the channel got moved into, and not every team a user is part of, for performance reasons.
+        // For every other team, we update the state here. Everything is correct on server side, but we update state here
+        // to avoid re-fetching all categories again for all teams.
+
+        const receivedChannel = action.data as Channel;
+        const newState: IDMappedObjects<ChannelCategory> = {};
+        const categoryIDs = Object.keys(state);
+
+        categoryIDs.forEach((categoryID) => {
+            if (categoryID.startsWith('channels_') && state[categoryID].team_id === receivedChannel.team_id && state[categoryID].channel_ids.indexOf(receivedChannel.id) < 0) {
+                // We don't need to worry about adding the channel in the right order as this is only
+                // an intermediate step, meant to handle the edge case of missing the upcoming "update category"
+                // websocket message, triggered on conversion of GM to private channel.
+                newState[categoryID] = {
+                    ...newState[categoryID],
+                    channel_ids: [...state[categoryID].channel_ids, receivedChannel.id],
+                };
+            } else {
+                newState[categoryID] = {
+                    ...state[categoryID],
+                    channel_ids: state[categoryID].channel_ids.filter((channelID) => channelID !== receivedChannel.id),
+                };
```

Comment:
This changes all categories, no matter if the channel was present or not. Not sure if that will be a "performance issue". Nevertheless, given the fact that this will not be called often, maybe it is ok to leave it like this.

Reply:
Yeah, not called often so okay IMHO.